hack/test-go.sh

```diff
@@ -1,129 +1,259 @@
+first_flag_arg=-1
+for (( i = 1; i < "$#"; i++ )); do
+    if [[ "${!i:0:1}" = "-" ]]; then
+        # we found an arg that begins with a dash, so we stop interpreting arguments 
+        # henceforth as packages and instead interpret them as flags to give to `go test`
+        first_flag_arg=$i
+        break
+    fi
+done
+
+package_args=
+if [[ ${first_flag_arg} -gt 0 ]]; then
+    # we recieved some `go test` flags as args, so we add them to the gotest_flags
+    gotest_flags="${gotest_flags} ${@:${first_flag_arg}}"
+
+    # the other args aside from $0 are packages to test
+    package_args="${@:1:${first_flag_arg}-1}"
 else
-  test_packages=`find_test_dirs "."; special_upstream_test_dirs`
+    # we got no flags as args, so all args are packages to test
+    package_args="$@"
 fi
 
-if [ -n "$PRINT_PACKAGES" ]; then
-  for package in $test_packages
-  do
-    echo $package
-  done
+# Determine packages to test
+test_packages=
+if [[ "${#package_args}" -gt 0 ]]; then
+    for package in ${package_args}; do
+        test_packages="${test_packages} ${OS_GO_PACKAGE}/${package}"
+    done
+else
+    # If no packages are given to test, we need to generate a list of all packages with unit tests
+    openshift_test_packages="$(list_test_packages_under '*')"  
+
+    kubernetes_path="Godeps/_workspace/src/k8s.io/kubernetes"
+    mandatory_kubernetes_packages="${OS_GO_PACKAGE}/${kubernetes_path}/pkg/api ${OS_GO_PACKAGE}/${kubernetes_path}/pkg/api/v1"
 
-  exit 0
+    test_packages="${openshift_test_packages} ${mandatory_kubernetes_packages}"
+
+    if [[ -n "${test_kube}" ]]; then
+        # we need to fine all of the kubernetes test suites, excluding those we directly whitelisted before, the end-to-end suite, and
+        # the go2idl tests which we currently do not support
+        optional_kubernetes_packages="$(find "${kubernetes_path}" -not \(                             \
+          \(                                                                                          \
+            -path "${kubernetes_path}/pkg/api"                                                        \
+            -o -path "${kubernetes_path}/pkg/api/v1"                                                  \
+            -o -path "${kubernetes_path}/test/e2e"                                                    \
+            -o -path "${kubernetes_path}/cmd/libs/go2idl/client-gen/testoutput/testgroup/unversioned" \
+          \) -prune                                                                                   \
+        \) -name '*_test.go' -print0 | xargs -0n1 dirname -z | sort -uz | xargs -0n1 printf "${OS_GO_PACKAGE}/%s\n")"
+
+        test_packages="${test_packages} ${optional_kubernetes_packages}"
+    fi
 fi
 
-export OPENSHIFT_ON_PANIC=crash
+if [[ -n "${dry_run}" ]]; then
+    echo "The following invocation of \`go test\` will be used by $0:"
+    echo "go test ${gotest_flags}"
```

Comment:
This fails to mention the `-coverprofile` flag that will be added if `${coverage_output_dir}` is used, but it's probably not worth adding an extra check here.